src/Simplex/Messaging/Agent.hs

```diff
@@ -1735,339 +1788,424 @@ processSMPTransmission c@AgentClient {smpClients, subQ} (tSess@(_, srv, _), v, s
-                  sq1' <- withStore' c $ \db -> setSndSwitchStatus db sq1 $ Just SSSendingQTEST
-                  let sqs' = updatedQs sq1' sqs
-                      conn' = DuplexConnection cData rqs sqs'
-                  notify . SWITCH QDSnd SPSecured $ connectionStats conn'
-                _ -> qError "QUSE: switching SndQueue not found in connection"
-            _ -> qError "QUSE: switched queue address not found in connection"
-
-        qError :: String -> m ()
-        qError = throwError . AGENT . A_QUEUE
-
-        smpInvitation :: ConnectionRequestUri 'CMInvitation -> ConnInfo -> m ()
-        smpInvitation connReq@(CRInvitationUri crData _) cInfo = do
-          logServer "<--" c srv rId "MSG <KEY>"
-          case conn of
-            ContactConnection {} -> do
-              g <- asks idsDrg
-              let newInv = NewInvitation {contactConnId = connId, connReq, recipientConnInfo = cInfo}
-              invId <- withStore c $ \db -> createInvitation db g newInv
-              let srvs = L.map qServer $ crSmpQueues crData
-              notify $ REQ invId srvs cInfo
-            _ -> prohibited
-
-        checkMsgIntegrity :: PrevExternalSndId -> ExternalSndId -> PrevRcvMsgHash -> ByteString -> MsgIntegrity
-        checkMsgIntegrity prevExtSndId extSndId internalPrevMsgHash receivedPrevMsgHash
-          | extSndId == prevExtSndId + 1 && internalPrevMsgHash == receivedPrevMsgHash = MsgOk
-          | extSndId < prevExtSndId = MsgError $ MsgBadId extSndId
-          | extSndId == prevExtSndId = MsgError MsgDuplicate -- ? deduplicate
-          | extSndId > prevExtSndId + 1 = MsgError $ MsgSkipped (prevExtSndId + 1) (extSndId - 1)
-          | internalPrevMsgHash /= receivedPrevMsgHash = MsgError MsgBadHash
-          | otherwise = MsgError MsgDuplicate -- this case is not possible
+          qDuplex :: String -> (Connection 'CDuplex -> m ()) -> m ()
+          qDuplex name a = case conn of
+            DuplexConnection {} -> a conn
+            _ -> qError $ name <> ": message must be sent to duplex connection"
+
+          newRatchetKey :: CR.E2ERatchetParams 'C.X448 -> m ()
+          newRatchetKey e2eOtherPartyParams@(CR.E2ERatchetParams e2eVersion k1Rcv k2Rcv) =
+            qDuplex "AgentRatchetKey" $ \duplexConn -> do
+              let rkHash = C.sha256Hash $ C.pubKeyBytes k1Rcv <> C.pubKeyBytes k2Rcv
+              unlessM (withStore' c $ \db -> checkProcessedRatchetKeyHashExists db connId rkHash) $ do
+                withStore' c $ \db -> addProcessedRatchetKeyHash db connId rkHash
+                AgentConfig {e2eEncryptVRange} <- asks config
+                unless (e2eVersion `isCompatible` e2eEncryptVRange) (throwError $ AGENT A_VERSION)
+                if rss == RSStarted
+                  then processReplyRatchetKey e2eEncryptVRange duplexConn
+                  else processInitiatingRatchetKey e2eEncryptVRange duplexConn
```

Comment:
these two functions are almost the same, and they will be even more similar if you don't delete ratchets in advise, so they can be refactored to one with some local conditions.